equinox/_filters.py

```diff
@@ -17,31 +27,40 @@
 #
 
 
-# Workaround https://github.com/patrick-kidger/equinox/issues/1095 in a backward
-# compatible manner.
-_NDARRAY_TYPES: Final = (np.ndarray, np.generic)
-_ARRAY_TYPES = (*_NDARRAY_TYPES, jax.Array)  # JAX < 0.7.2
+# Define array types
+# Note that we don't accept `np.generic` as a suitable duck-type for arrays,
+# because `generic` includes `np.object_` and `np.flexible` which are not
+# array-like.
+NDArrayType: TypeAlias = np.ndarray | np.number | np.bool_
+_NDARRAY_TYPES: Final = get_args(NDArrayType)
+_ARRAY_TYPES = _NDARRAY_TYPES + (jax.Array,)
```

Comment:
I think the change to type narrow `NDArrayType` from `generic` to `number | bool` means it excludes `bfloat16`!
🤔 
We want to include that, while still excluding `np.object_` and `np.flexible`.

Reply:
Ah, interesting! I think just adding `jax.dtypes.bfloat16` should work?

It would also probably be fine to include `np.generic` as an overestimation, to be robust to some of the more esoteric 4-bit (etc) dtypes as well?

Reply:
That looks to be necessary since jax doesn't provide a nice union of the ml_dtypes that it imports.
I wish they would! Or just subclass `np.number`